rcl/src/rcl/type_version_hash.c

```diff
@@ -0,0 +1,256 @@
+// Copyright 2023 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <yaml.h>
+
+#include "rcl/allocator.h"
+#include "rcl/error_handling.h"
+#include "rcl/type_version_hash.h"
+#include "rcutils/types/char_array.h"
+#include "rcutils/sha256.h"
+#include "type_description_interfaces/msg/type_description.h"
+
+static int yaml_write_handler(void * ext, uint8_t * buffer, size_t size)
+{
+  rcutils_char_array_t * repr = (rcutils_char_array_t *)ext;
+  rcutils_ret_t res = rcutils_char_array_strncat(repr, (char *)buffer, size);
+  return res == RCL_RET_OK ? 1 : 0;
+}
+
+static inline int start_sequence(yaml_emitter_t * emitter)
+{
+  yaml_event_t event;
+  return
+    yaml_sequence_start_event_initialize(&event, NULL, NULL, 1, YAML_FLOW_SEQUENCE_STYLE) &&
+    yaml_emitter_emit(emitter, &event);
+}
+
+static inline int end_sequence(yaml_emitter_t * emitter)
+{
+  yaml_event_t event;
+  return
+    yaml_sequence_end_event_initialize(&event) &&
+    yaml_emitter_emit(emitter, &event);
+}
+
+static inline int start_mapping(yaml_emitter_t * emitter)
+{
+  yaml_event_t event;
+  return
+    yaml_mapping_start_event_initialize(&event, NULL, NULL, 1, YAML_FLOW_MAPPING_STYLE) &&
+    yaml_emitter_emit(emitter, &event);
+}
+
+static inline int end_mapping(yaml_emitter_t * emitter)
+{
+  yaml_event_t event;
+  return
+    yaml_mapping_end_event_initialize(&event) &&
+    yaml_emitter_emit(emitter, &event);
+}
+
+static int emit_key(yaml_emitter_t * emitter, const char * key)
+{
+  yaml_event_t event;
+  return
+    yaml_scalar_event_initialize(
+    &event, NULL, NULL, (yaml_char_t *)key, strlen(key), 0, 1, YAML_DOUBLE_QUOTED_SCALAR_STYLE) &&
+    yaml_emitter_emit(emitter, &event);
+}
+
+static int emit_int(yaml_emitter_t * emitter, size_t val, const char * fmt)
+{
+  // longest uint64 is 21 decimal digits
```

Comment:
More accurately, the longest uint64 is 20 decimal digits, with one byte for the trailing `\0`.